src/Module.Server/Rewards/CrpgRewardServer.cs

```diff
@@ -199,13 +243,141 @@ public override void OnRemoveBehavior()
         }
         catch (Exception e)
         {
             Debug.Print("Couldn't update users: " + e);
             SendErrorToPeers(crpgPeerByUserId);
         }
+
+        _agentsThatGotHitThisRoundByCrpgUserId.Clear();
+    }
+
+    private static void RegisterHitForAffectedCrpgAgent(CrpgAgent affectedCrpgAgent, int affectorCrpgUserId, int inflictedDamage, bool isTeamHit)
+    {
+        int damageDone = Math.Min(inflictedDamage, affectedCrpgAgent.CurrentHealth);
+        if (affectedCrpgAgent.Hitters.TryGetValue(affectorCrpgUserId, out CrpgAgent.Hitter? affectedCrpgAgentAttacker) && affectedCrpgAgentAttacker != null)
+        {
+            affectedCrpgAgentAttacker.TotalDamageDone += damageDone;
+        }
+        else
+        {
+            affectedCrpgAgent.Hitters.Add(affectorCrpgUserId, new CrpgAgent.Hitter
+            {
+                CharacterId = affectorCrpgUserId,
+                TotalDamageDone = Math.Min(inflictedDamage, affectedCrpgAgent.CurrentHealth),
+                IsSameTeam = isTeamHit,
+            });
+        }
+
+        affectedCrpgAgent.CurrentHealth -= damageDone;
+    }
+
+    private bool IsMissionGameTypeBattle(Mission mission)
+    {
+        var missionMultiplayerGameModeBaseClient = mission.GetMissionBehavior<MissionMultiplayerGameModeBaseClient>();
+        if (missionMultiplayerGameModeBaseClient == null)
+        {
+            return false;
+        }
+
+        return missionMultiplayerGameModeBaseClient.GameType == MissionLobbyComponent.MultiplayerGameType.Battle;
+    }
+
+    private bool TryGetCrpgUserIdSafely(Agent agent, out int crpgUserId)
+    {
+        crpgUserId = 0;
+        if (agent.MissionPeer == null)
+        {
+            return false;
+        }
+
+        if (agent.MissionPeer.GetNetworkPeer() == null)
+        {
+            return false;
+        }
+
+        var crpgPeer = agent.MissionPeer.GetNetworkPeer().GetComponent<CrpgPeer>();
+        if (crpgPeer == null)
+        {
+            return false;
+        }
+
+        if (crpgPeer.User == null)
+        {
+            return false;
+        }
+
+        crpgUserId = crpgPeer.User.Id;
+        return true;
+    }
+
+    private Dictionary<int, int> CalculateNetCompensationByCrpgUserId(Dictionary<int, int> repairCostByCrpgUserId)
+    {
+        Dictionary<int, int> netCompensationByCrpgUserId = new();
+        foreach (var affectedEntry in _agentsThatGotHitThisRoundByCrpgUserId)
+        {
+            int affectedCrpgUserId = affectedEntry.Key;
+            CrpgAgent affectedCrpgAgent = affectedEntry.Value;
+            foreach (var hitterEntry in affectedCrpgAgent.Hitters)
+            {
+                int attackerCrpgUserId = hitterEntry.Key;
+                Hitter affectedCrpgAgentAttacker = hitterEntry.Value;
+                if (!affectedCrpgAgentAttacker.IsSameTeam)

```

Comment:
Are we not only interested by team-hits?

Reply:
as answered above

Reply:
Because i gather all hits, i need to ignore the hits that are not from the same team.